src/core/init/stall_avoider.ts

```diff
@@ -0,0 +1,357 @@
+          if (realSeekTime <= mediaElement.currentTime) {
+            log.info("Init: position to seek already reached, no seeking",
+                     mediaElement.currentTime, realSeekTime);
+          } else {
+            log.warn("SA: skippable discontinuity found in the stream",
+                     position, realSeekTime);
+            mediaElement.currentTime = realSeekTime;
+            return EVENTS.warning(generateDiscontinuityError(stalledPosition,
+                                                             realSeekTime));
+          }
+        }
+      }
+
+      // Is it a browser bug? -> force seek at the same current time
+      if (isPlaybackStuck(position,
+                          currentRange,
+                          state,
+                          stalled !== null)
+      ) {
+        log.warn("Init: After freeze seek", position, currentRange);
+        mediaElement.currentTime = position;
+        return EVENTS.warning(generateDiscontinuityError(position,
+                                                         position));
+
+      }
+
+      const freezePosition = stalledPosition ?? position;
+
+      // Is it a very short discontinuity in buffer ? -> Seek at the beginning of the
+      //                                                 next range
+      //
+      // Discontinuity check in case we are close a buffered range but still
+      // calculate a stalled state. This is useful for some
+      // implementation that might drop an injected segment, or in
+      // case of small discontinuity in the content.
+      const nextBufferRangeGap = getNextRangeGap(buffered, freezePosition);
+      if (nextBufferRangeGap < BUFFER_DISCONTINUITY_THRESHOLD) {
+        const seekTo = (freezePosition + nextBufferRangeGap + EPSILON);
+        if (mediaElement.currentTime < seekTo) {
+          log.warn("Init: discontinuity encountered inferior to the threshold",
+                   freezePosition, seekTo, BUFFER_DISCONTINUITY_THRESHOLD);
+          mediaElement.currentTime = seekTo;
+          return EVENTS.warning(generateDiscontinuityError(freezePosition, seekTo));
+        }
+      }
+
+      // Are we in a discontinuity between periods ? -> Seek at the beginning of the
+      //                                                next period
+      for (let i = 0; i < manifest.periods.length; i++) {
+        const period = manifest.periods[i];
+        if (period.end !== undefined && period.end <= freezePosition) {
+          if (i + 1 < manifest.periods.length &&
+              manifest.periods[i + 1].start > freezePosition &&
+              manifest.periods[i + 1].start > mediaElement.currentTime)
+          {
+            const nextPeriod = manifest.periods[i + 1];
+            mediaElement.currentTime = nextPeriod.start;
+            return EVENTS.warning(generateDiscontinuityError(freezePosition,
+                                                             nextPeriod.start));
+
+          }
+          break;
```

Comment:
It seems that this `break` will avoid browsing through the array if the upper condition is not met with the first period.

Proposition :
Turning line 203 into :       `for (let i = manifest.periods.length - 1; i >= 0; i--) {`

Reply:
Yes, there is an issue with that logic...

I tried to fix it.